personalrobotics/ada_feeding
# Add SegmentFromPoint ROS Node

## What does this PR do?

This PR implements a ROSw node that creates an action server for food segmentation using [Segment Anything](https://segment-anything.com/). Specifically, this node does the following:

1. Checks if the model is there and, if not, downloads it.
2. Loads the model onto `cuda` if available, else `cpu`.
3. Launches an action server that:
    1. Takes in a user-specified point.
    2. Gets the latest image from the robot's camera.
    3. Runs Segment Anything on the image with the specified point.
    4. Cleans up and returns the top 3 (can be modified by ROS param) masks. Currently, "cleaning" only consists of removing any pixel that is not part of the connected component containing the seed point.
    5. **NOTE** that this action server _**does not**_ return any feedback. I tried [an implementation](https://github.com/personalrobotics/ada_feeding/blob/8ce3d381e1f5f69e466b1c59cb572f1757f5510c/ada_feeding_perception/ada_feeding_perception/segment_from_point.py) that launched the segmentation function in another thread and returned feedback, and it took 4x as long.

This PR includes the commits from #22 .

## How was this PR tested?

1. Build the workspace: `colcon build`
2. Source the workspace: `source install/setup.bash`
3. Run the action server(s): `ros2 launch ada_feeding_perception ada_feeding_perception_launch.xml`
4. Check that Segment Anything is working on your computer:
    - [x] `cd cd ada_feeding_perception`
    - [x] `python3 test_sam.py --input_image ../test/food_img/sete_00100.jpg --input_point [800,500]`
6. Test with web app on a CPU machine (4 cores, 8 GB RAM).
    - [x] Launch the dummy motion and RealSense nodes: `ros2 launch feeding_web_app_ros2_test feeding_web_app_dummy_nodes_launch.xml run_perception:=false` (this requires the code from [feeding_web_interface#58](https://github.com/personalrobotics/feeding_web_interface/pull/58))
    - [x] Ensure that food segmentation works as expected:
    - [x] Ensure that clicking once segments the image as expected. It should compute the masks w.r.t. the image that is visible when you clicked.
    - [x] Ensure that clicking a second time while the previous goal is running does not override the first goal, and instead cancels the second goal.
    - [x] Ensure that the computed masks make sense, for several different images and points.
7. Test with web app on a GPU machine (LoveLace).
    - [x] Launch the dummy motion and RealSense nodes: `ros2 launch feeding_web_app_ros2_test feeding_web_app_dummy_nodes_launch.xml run_perception:=false` (this requires the code from [feeding_web_interface#58](https://github.com/personalrobotics/feeding_web_interface/pull/58))
    - [x] Ensure that food segmentation works as expected:
    - [x] Ensure that clicking once segments the image as expected. It should compute the masks w.r.t. the image that is visible when you clicked.
    - [x] Ensure that clicking a second time while the previous goal is running does not override the first goal, and instead cancels the second goal.
    - [x] Ensure that the computed masks make sense, for several different images and points.
8. Visually test the action server in isolation: 
    - [x] Online mode: `ros2 launch ada_feeding_perception test_segment_from_point_launch.xml`. Click a point on the live stream, and see the segmented results, and ensure they make sense. Note that due to delays from rendering this GUI, this sometimes took ~90s on my CPU machine.
    - [x] Offline mode: (1) In `config/test_segment_from_point.yaml` change the `mode` parameter to `offline`; (2) `ros2 launch ada_feeding_perception test_segment_from_point_launch.xml`; (3) visually inspect the results in `<path/to/your/workspace>/install/ada_feeding_perception/share/ada_feeding_perception/test_img/output`

# Rough Speed Benchmarking
A very rough speed benchmarking (where I started the timer when I clicked the image on the web app and stopped the timer when the buttons with masks appeared) revealed the following:
- On LoveLace's GPU, it takes ~0.6s for masks to appear.
- On LoveLace's CPU it takes ~4s for masks to appear.
- On Amal's Ubuntu VM (4 core, 8GB RAM) CPU, it takes ~ 15s to appear.

# Before opening a pull request
- [X] Run your code through [pylint](https://pylint.readthedocs.io/en/latest/) and address (most) warnings/errors: `pylint --recursive=y .`

# Before Merging
